File: IBCMAndroid/app/src/main/java/com/example/ibcmserver_init/data/repository/EventRepository.kt
package com.example.ibcmserver_init.data.repository

import com.example.ibcmserver_init.data.local.EventDao
import com.example.ibcmserver_init.data.model.Event
import com.example.ibcmserver_init.data.model.EventRequest
import com.example.ibcmserver_init.data.model.Review
import com.example.ibcmserver_init.data.remote.EventApi
import com.example.ibcmserver_init.utils.NetworkUtils
import com.example.ibcmserver_init.utils.Resource
import com.example.ibcmserver_init.utils.networkBoundResource
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import javax.inject.Inject
import java.io.IOException
import java.time.LocalDate
import java.time.LocalTime
import com.example.ibcmserver_init.data.api.*
import com.example.ibcmserver_init.ui.screens.event.*

interface EventRepository {
    suspend fun createEvent(event: Event): Result<Event>
    suspend fun getEventById(eventId: String): Event
    suspend fun updateEvent(event: Event): Result<Event>
    suspend fun deleteEvent(eventId: String): Result<Unit>
    suspend fun searchEvents(query: String): List<Event>
    suspend fun getEventsByCategory(category: String): List<Event>
    suspend fun getEventsByUser(userId: String): List<Event>
    suspend fun joinEvent(eventId: String): Result<Event>
    suspend fun leaveEvent(eventId: String): Result<Event>
    suspend fun addEventReview(eventId: String, review: Review): Result<Review>
    suspend fun addComment(eventId: String, text: String): Event
    suspend fun addCommentWithImage(eventId: String, text: String, imageUrl: String): Event
    suspend fun deleteComment(eventId: String, commentId: String): Event
    suspend fun setReminder(eventId: String, enabled: Boolean): Event
    suspend fun upgradeEventVisibility(eventId: String): Result<Event>
    fun observeEvent(eventId: String): Flow<Event>
    fun observeEventsByCategory(category: String): Flow<List<Event>>
    fun observeEventsByUser(userId: String): Flow<List<Event>>

    suspend fun getEvent(eventId: String): Event

    suspend fun searchEvents(creatorId: String? = null, attendeeId: String? = null): List<Event>

    suspend fun getNearbyEvents(
        latitude: Double,
        longitude: Double,
        radius: Double,
        category: String? = null
    ): List<Event>

    suspend fun getUpcomingEvents(): List<Event>

    suspend fun getPopularEvents(): List<Event>

    suspend fun getEventRequests(eventId: String): Flow<List<EventRequest>>

    suspend fun createEventRequest(request: EventRequest): Result<EventRequest>

    suspend fun updateEventRequest(request: EventRequest): Result<EventRequest>

    suspend fun getEventReviews(eventId: String): Flow<List<Review>>

    suspend fun getRating(eventId: String): Float

    suspend fun rateEvent(eventId: String, rating: Float, review: String?): Event

    suspend fun getEventUpdates(eventId: String): Flow<Event>

    suspend fun getTrendingEvents(category: String? = null): List<Event>

    suspend fun generateEventDescription(request: DescriptionGenerationRequest): Flow<Result<DescriptionGenerationResponse>>
    suspend fun processEventImages(images: List<String>): Flow<Result<ImageProcessingResponse>>
    suspend fun processEventVideo(videoPath: String): Flow<Result<VideoProcessingResponse>>
}

class EventRepository @Inject constructor(
    private val eventApi: EventApi,
    private val eventDao: EventDao,
    private val networkUtils: NetworkUtils
) {
    fun getNearbyEventsWithCache(
        latitude: Double,
        longitude: Double,
        radius: Double,
        category: String? = null,
        forceRefresh: Boolean = false
    ): Flow<Resource<List<Event>>> = networkBoundResource(
        query = {
            eventDao.getNearbyEvents(latitude, longitude, radius, category)
        },
        fetch = {
            eventApi.getNearbyEvents(latitude, longitude, radius, category)
        },
        saveFetchResult = { events ->
            eventDao.insertEvents(events)
        },
        shouldFetch = { cachedEvents ->
            forceRefresh || cachedEvents.isEmpty() || !networkUtils.isNetworkAvailable()
        }
    )

    fun getTrendingEventsWithCache(
        category: String? = null,
        forceRefresh: Boolean = false
    ): Flow<Resource<List<Event>>> = networkBoundResource(
        query = {
            eventDao.getTrendingEvents(category)
        },
        fetch = {
            eventApi.getTrendingEvents(category)
        },
        saveFetchResult = { events ->
            eventDao.insertEvents(events)
        },
        shouldFetch = { cachedEvents ->
            forceRefresh || cachedEvents.isEmpty() || !networkUtils.isNetworkAvailable()
        }
    )

    suspend fun getNearbyEvents(
        latitude: Double,
        longitude: Double,
        radius: Double,
        category: String? = null
    ): List<Event> {
        if (!networkUtils.isNetworkAvailable()) {
            return eventDao.getNearbyEventsSync(latitude, longitude, radius, category)
        }
        return try {
            val events = eventApi.getNearbyEvents(latitude, longitude, radius, category)
            eventDao.insertEvents(events)
            events
        } catch (e: Exception) {
            eventDao.getNearbyEventsSync(latitude, longitude, radius, category)
        }
    }

    suspend fun getTrendingEvents(category: String? = null): List<Event> {
        if (!networkUtils.isNetworkAvailable()) {
            return eventDao.getTrendingEventsSync(category)
        }
        return try {
            val events = eventApi.getTrendingEvents(category)
            eventDao.insertEvents(events)
            events
        } catch (e: Exception) {
            eventDao.getTrendingEventsSync(category)
        }
    }

    suspend fun searchEvents(query: String, filters: Map<String, String>): List<Event> {
        if (!networkUtils.isNetworkAvailable()) {
            return eventDao.searchEvents(query, filters)
        }
        return try {
            val events = eventApi.searchEvents(query, filters)
            eventDao.insertEvents(events)
            events
        } catch (e: Exception) {
            eventDao.searchEvents(query, filters)
        }
    }

    suspend fun getEventById(eventId: String): Event? {
        return try {
            if (networkUtils.isNetworkAvailable()) {
                val event = eventApi.getEventById(eventId)
                eventDao.insertEvent(event)
                event
            } else {
                eventDao.getEventById(eventId)
            }
        } catch (e: Exception) {
            eventDao.getEventById(eventId)
        }
    }

    suspend fun createEvent(event: Event): Event {
        if (!networkUtils.isNetworkAvailable()) {
            throw IOException("No network connection available")
        }
        return eventApi.createEvent(event).also {
            eventDao.insertEvent(it)
        }
    }

    suspend fun updateEvent(eventId: String, event: Event): Event {
        if (!networkUtils.isNetworkAvailable()) {
            throw IOException("No network connection available")
        }
        return eventApi.updateEvent(eventId, event).also {
            eventDao.insertEvent(it)
        }
    }

    suspend fun deleteEvent(eventId: String) {
        if (!networkUtils.isNetworkAvailable()) {
            throw IOException("No network connection available")
        }
        eventApi.deleteEvent(eventId)
        eventDao.deleteEvent(eventId)
    }

    suspend fun createEvent(request: CreateEventRequest): Flow<Result<CreateEventResponse>> = flow {
        try {
            val response = eventApi.createEvent(request)
            if (response.isSuccessful) {
                response.body()?.let { emit(Result.success(it)) }
            } else {
                emit(Result.failure(Exception("Failed to create event: ${response.message()}")))
            }
        } catch (e: Exception) {
            emit(Result.failure(e))
        }
    }

    suspend fun getEventOptimizations(request: EventOptimizationRequest): Flow<Result<EventOptimizations>> = flow {
        try {
            val optimizations = eventApi.getEventOptimizations(request)
            emit(Result.success(optimizations))
        } catch (e: Exception) {
            emit(Result.failure(e))
        }
    }

    suspend fun setupEventAutomation(eventId: String, request: AutomationRequest): Flow<Result<Unit>> = flow {
        try {
            val response = eventApi.setupEventAutomation(eventId, request)
            if (response.isSuccessful) {
                emit(Result.success(Unit))
            } else {
                emit(Result.failure(Exception("Failed to setup automation: ${response.message()}")))
            }
        } catch (e: Exception) {
            emit(Result.failure(e))
        }
    }

    suspend fun autoGenerateEvent(request: AutoGenerateEventRequest): Flow<Result<CreateEventResponse>> = flow {
        try {
            val response = eventApi.autoGenerateEvent(request)
            if (response.isSuccessful) {
                response.body()?.let { emit(Result.success(it)) }
            } else {
                emit(Result.failure(Exception("Failed to auto-generate event: ${response.message()}")))
            }
        } catch (e: Exception) {
            emit(Result.failure(e))
        }
    }

    override suspend fun generateEventDescription(
        request: DescriptionGenerationRequest
    ): Flow<Result<DescriptionGenerationResponse>> = flow {
        try {
            val response = eventApi.generateEventDescription(request)
            if (response.isSuccessful) {
                response.body()?.let { emit(Result.success(it)) }
            } else {
                emit(Result.failure(Exception("Failed to generate description: ${response.message()}")))
            }
        } catch (e: Exception) {
            emit(Result.failure(e))
        }
    }

    override suspend fun processEventImages(
        images: List<String>
    ): Flow<Result<ImageProcessingResponse>> = flow {
        try {
            val response = eventApi.processEventImages(images)
            if (response.isSuccessful) {
                response.body()?.let { emit(Result.success(it)) }
            } else {
                emit(Result.failure(Exception("Failed to process images: ${response.message()}")))
            }
        } catch (e: Exception) {
            emit(Result.failure(e))
        }
    }

    override suspend fun processEventVideo(
        videoPath: String
    ): Flow<Result<VideoProcessingResponse>> = flow {
        try {
            val response = eventApi.processEventVideo(videoPath)
            if (response.isSuccessful) {
                response.body()?.let { emit(Result.success(it)) }
            } else {
                emit(Result.failure(Exception("Failed to process video: ${response.message()}")))
            }
        } catch (e: Exception) {
            emit(Result.failure(e))
        }
    }
} 

File: IBCM-ios/IBCM/IBCM/Repositories/EventRepository.swift
import Foundation

protocol EventRepository {
    func createEvent(event: Event) async throws -> Event
    func getEventById(eventId: String) async throws -> Event
    func updateEvent(event: Event) async throws -> Event
    func deleteEvent(eventId: String) async throws
    func searchEvents(query: String) async throws -> [Event]
    func getEventsByCategory(category: String) async throws -> [Event]
    func getEventsByUser(userId: String) async throws -> [Event]
    func joinEvent(eventId: String) async throws -> Event
    func leaveEvent(eventId: String) async throws -> Event
    func addEventReview(eventId: String, review: Review) async throws -> Review
    func addComment(eventId: String, text: String) async throws -> Event
    func addCommentWithImage(eventId: String, text: String, imageUrl: String) async throws -> Event
    func deleteComment(eventId: String, commentId: String) async throws -> Event
    func setReminder(eventId: String, enabled: Bool) async throws -> Event
    func upgradeEventVisibility(eventId: String) async throws -> Event
    func observeEvent(eventId: String) -> AsyncStream<Event>
    func observeEventsByCategory(category: String) -> AsyncStream<[Event]>
    func observeEventsByUser(userId: String) -> AsyncStream<[Event]>
    func getNearbyEvents(latitude: Double, longitude: Double, radius: Double) async throws -> [Event]
    func getUpcomingEvents() async throws -> [Event]
    func getPopularEvents() async throws -> [Event]
    func getEventRequests(eventId: String) async throws -> [EventRequest]
    func createEventRequest(request: EventRequest) async throws -> EventRequest
    func updateEventRequest(request: EventRequest) async throws -> EventRequest
    func getEventReviews(eventId: String) async throws -> [Review]
    func getRating(eventId: String) async throws -> Float
    func rateEvent(eventId: String, rating: Float, review: String?) async throws -> Event
    func getEventUpdates(eventId: String) async throws -> Event
}

class EventRepositoryImpl: EventRepository {
    private let apiService: APIService
    private let webSocket: WebSocketService
    private let cache: NSCache<NSString, CachedEvent>
    private var eventObservers: [String: AsyncStream<Event>] = [:]
    private var categoryObservers: [String: AsyncStream<[Event]>] = [:]
    private var userObservers: [String: AsyncStream<[Event]>] = [:]
    
    init(apiService: APIService = .shared, webSocket: WebSocketService = .shared) {
        self.apiService = apiService
        self.webSocket = webSocket
        self.cache = NSCache()
    }
    
    func createEvent(event: Event) async throws -> Event {
        let response: EventResponse = try await apiService.request(
            endpoint: "/events",
            method: "POST",
            body: try JSONEncoder().encode(event)
        )
        
        let createdEvent = response.data
        cache.setObject(CachedEvent(event: createdEvent, timestamp: Date()), forKey: createdEvent.id as NSString)
        return createdEvent
    }
    
    func getEventById(eventId: String) async throws -> Event {
        if let cached = cache.object(forKey: eventId as NSString) {
            if Date().timeIntervalSince(cached.timestamp) < 300 { // 5 minutes cache
                return cached.event
            }
        }
        
        let response: EventResponse = try await apiService.request(
            endpoint: "/events/\(eventId)",
            method: "GET"
        )
        
        let event = response.data
        cache.setObject(CachedEvent(event: event, timestamp: Date()), forKey: event.id as NSString)
        return event
    }
    
    func updateEvent(event: Event) async throws -> Event {
        let response: EventResponse = try await apiService.request(
            endpoint: "/events/\(event.id)",
            method: "PUT",
            body: try JSONEncoder().encode(event)
        )
        
        let updatedEvent = response.data
        cache.setObject(CachedEvent(event: updatedEvent, timestamp: Date()), forKey: updatedEvent.id as NSString)
        return updatedEvent
    }
    
    func deleteEvent(eventId: String) async throws {
        let response: BasicResponse = try await apiService.request(
            endpoint: "/events/\(eventId)",
            method: "DELETE"
        )
        
        if response.success {
            cache.removeObject(forKey: eventId as NSString)
        } else {
            throw EventError.deleteFailed
        }
    }
    
    func searchEvents(query: String) async throws -> [Event] {
        let response: EventListResponse = try await apiService.request(
            endpoint: "/events/search",
            method: "GET",
            queryItems: [URLQueryItem(name: "query", value: query)]
        )
        
        response.data.forEach { event in
            cache.setObject(CachedEvent(event: event, timestamp: Date()), forKey: event.id as NSString)
        }
        
        return response.data
    }
    
    func getEventsByCategory(category: String) async throws -> [Event] {
        let response: EventListResponse = try await apiService.request(
            endpoint: "/events/category/\(category)",
            method: "GET"
        )
        
        response.data.forEach { event in
            cache.setObject(CachedEvent(event: event, timestamp: Date()), forKey: event.id as NSString)
        }
        
        return response.data
    }
    
    func getEventsByUser(userId: String) async throws -> [Event] {
        let response: EventListResponse = try await apiService.request(
            endpoint: "/events/user/\(userId)",
            method: "GET"
        )
        
        response.data.forEach { event in
            cache.setObject(CachedEvent(event: event, timestamp: Date()), forKey: event.id as NSString)
        }
        
        return response.data
    }
    
    func joinEvent(eventId: String) async throws -> Event {
        let response: EventResponse = try await apiService.request(
            endpoint: "/events/\(eventId)/join",
            method: "POST"
        )
        
        let event = response.data
        cache.setObject(CachedEvent(event: event, timestamp: Date()), forKey: event.id as NSString)
        return event
    }
    
    func leaveEvent(eventId: String) async throws -> Event {
        let response: EventResponse = try await apiService.request(
            endpoint: "/events/\(eventId)/leave",
            method: "POST"
        )
        
        let event = response.data
        cache.setObject(CachedEvent(event: event, timestamp: Date()), forKey: event.id as NSString)
        return event
    }
    
    func addEventReview(eventId: String, review: Review) async throws -> Review {
        let response: ReviewResponse = try await apiService.request(
            endpoint: "/events/\(eventId)/reviews",
            method: "POST",
            body: try JSONEncoder().encode(review)
        )
        return response.data
    }
    
    func addComment(eventId: String, text: String) async throws -> Event {
        let response: EventResponse = try await apiService.request(
            endpoint: "/events/\(eventId)/comments",
            method: "POST",
            body: try JSONEncoder().encode(["text": text])
        )
        
        let event = response.data
        cache.setObject(CachedEvent(event: event, timestamp: Date()), forKey: event.id as NSString)
        return event
    }
    
    func addCommentWithImage(eventId: String, text: String, imageUrl: String) async throws -> Event {
        let response: EventResponse = try await apiService.request(
            endpoint: "/events/\(eventId)/comments",
            method: "POST",
            body: try JSONEncoder().encode([
                "text": text,
                "imageUrl": imageUrl
            ])
        )
        
        let event = response.data
        cache.setObject(CachedEvent(event: event, timestamp: Date()), forKey: event.id as NSString)
        return event
    }
    
    func deleteComment(eventId: String, commentId: String) async throws -> Event {
        let response: EventResponse = try await apiService.request(
            endpoint: "/events/\(eventId)/comments/\(commentId)",
            method: "DELETE"
        )
        
        let event = response.data
        cache.setObject(CachedEvent(event: event, timestamp: Date()), forKey: event.id as NSString)
        return event
    }
    
    func setReminder(eventId: String, enabled: Bool) async throws -> Event {
        let response: EventResponse = try await apiService.request(
            endpoint: "/events/\(eventId)/reminder",
            method: "PUT",
            body: try JSONEncoder().encode(["enabled": enabled])
        )
        
        let event = response.data
        cache.setObject(CachedEvent(event: event, timestamp: Date()), forKey: event.id as NSString)
        return event
    }
    
    func upgradeEventVisibility(eventId: String) async throws -> Event {
        let response: EventResponse = try await apiService.request(
            endpoint: "/events/\(eventId)/upgrade",
            method: "POST"
        )
        
        let event = response.data
        cache.setObject(CachedEvent(event: event, timestamp: Date()), forKey: event.id as NSString)
        return event
    }
    
    func observeEvent(eventId: String) -> AsyncStream<Event> {
        if let existing = eventObservers[eventId] {
            return existing
        }
        
        let stream = AsyncStream<Event> { continuation in
            Task {
                for await message in webSocket.subscribe(to: "event.\(eventId)") {
                    if let event = try? JSONDecoder().decode(Event.self, from: message) {
                        continuation.yield(event)
                    }
                }
            }
        }
        
        eventObservers[eventId] = stream
        return stream
    }
    
    func observeEventsByCategory(category: String) -> AsyncStream<[Event]> {
        if let existing = categoryObservers[category] {
            return existing
        }
        
        let stream = AsyncStream<[Event]> { continuation in
            Task {
                for await message in webSocket.subscribe(to: "events.category.\(category)") {
                    if let events = try? JSONDecoder().decode([Event].self, from: message) {
                        continuation.yield(events)
                    }
                }
            }
        }
        
        categoryObservers[category] = stream
        return stream
    }
    
    func observeEventsByUser(userId: String) -> AsyncStream<[Event]> {
        if let existing = userObservers[userId] {
            return existing
        }
        
        let stream = AsyncStream<[Event]> { continuation in
            Task {
                for await message in webSocket.subscribe(to: "events.user.\(userId)") {
                    if let events = try? JSONDecoder().decode([Event].self, from: message) {
                        continuation.yield(events)
                    }
                }
            }
        }
        
        userObservers[userId] = stream
        return stream
    }
    
    func getNearbyEvents(latitude: Double, longitude: Double, radius: Double) async throws -> [Event] {
        let response: EventListResponse = try await apiService.request(
            endpoint: "/events/nearby",
            method: "GET",
            queryItems: [
                URLQueryItem(name: "latitude", value: "\(latitude)"),
                URLQueryItem(name: "longitude", value: "\(longitude)"),
                URLQueryItem(name: "radius", value: "\(radius)")
            ]
        )
        return response.data
    }
    
    func getUpcomingEvents() async throws -> [Event] {
        let response: EventListResponse = try await apiService.request(
            endpoint: "/events/upcoming",
            method: "GET"
        )
        return response.data
    }
    
    func getPopularEvents() async throws -> [Event] {
        let response: EventListResponse = try await apiService.request(
            endpoint: "/events/popular",
            method: "GET"
        )
        return response.data
    }
    
    func getEventRequests(eventId: String) async throws -> [EventRequest] {
        let response: EventRequestListResponse = try await apiService.request(
            endpoint: "/events/\(eventId)/requests",
            method: "GET"
        )
        return response.data
    }
    
    func createEventRequest(request: EventRequest) async throws -> EventRequest {
        let response: EventRequestResponse = try await apiService.request(
            endpoint: "/events/requests",
            method: "POST",
            body: try JSONEncoder().encode(request)
        )
        return response.data
    }
    
    func updateEventRequest(request: EventRequest) async throws -> EventRequest {
        let response: EventRequestResponse = try await apiService.request(
            endpoint: "/events/requests/\(request.id)",
            method: "PUT",
            body: try JSONEncoder().encode(request)
        )
        return response.data
    }
    
    func getEventReviews(eventId: String) async throws -> [Review] {
        let response: ReviewListResponse = try await apiService.request(
            endpoint: "/events/\(eventId)/reviews",
            method: "GET"
        )
        return response.data
    }
    
    func getRating(eventId: String) async throws -> Float {
        let response: RatingResponse = try await apiService.request(
            endpoint: "/events/\(eventId)/rating",
            method: "GET"
        )
        return response.rating
    }
    
    func rateEvent(eventId: String, rating: Float, review: String?) async throws -> Event {
        let response: EventResponse = try await apiService.request(
            endpoint: "/events/\(eventId)/rate",
            method: "POST",
            body: try JSONEncoder().encode([
                "rating": rating,
                "review": review
            ])
        )
        
        let event = response.data
        cache.setObject(CachedEvent(event: event, timestamp: Date()), forKey: event.id as NSString)
        return event
    }
    
    func getEventUpdates(eventId: String) async throws -> Event {
        let response: EventResponse = try await apiService.request(
            endpoint: "/events/\(eventId)/updates",
            method: "GET"
        )
        
        let event = response.data
        cache.setObject(CachedEvent(event: event, timestamp: Date()), forKey: event.id as NSString)
        return event
    }
}

// MARK: - Cache Types
private class CachedEvent {
    let event: Event
    let timestamp: Date
    
    init(event: Event, timestamp: Date) {
        self.event = event
        self.timestamp = timestamp
    }
}

// MARK: - Response Types
struct EventResponse: Codable {
    let success: Bool
    let data: Event
    let message: String?
}

struct EventListResponse: Codable {
    let success: Bool
    let data: [Event]
    let message: String?
}

struct ReviewResponse: Codable {
    let success: Bool
    let data: Review
    let message: String?
}

struct ReviewListResponse: Codable {
    let success: Bool
    let data: [Review]
    let message: String?
}

struct EventRequestResponse: Codable {
    let success: Bool
    let data: EventRequest
    let message: String?
}

struct EventRequestListResponse: Codable {
    let success: Bool
    let data: [EventRequest]
    let message: String?
}

struct RatingResponse: Codable {
    let success: Bool
    let rating: Float
    let message: String?
}

// MARK: - Errors
enum EventError: LocalizedError {
    case invalidEvent
    case eventNotFound
    case createFailed
    case updateFailed
    case deleteFailed
    case joinFailed
    case leaveFailed
    case reviewFailed
    case commentFailed
    case upgradeFailed
    
    var errorDescription: String? {
        switch self {
        case .invalidEvent:
            return "Invalid event"
        case .eventNotFound:
            return "Event not found"
        case .createFailed:
            return "Failed to create event"
        case .updateFailed:
            return "Failed to update event"
        case .deleteFailed:
            return "Failed to delete event"
        case .joinFailed:
            return "Failed to join event"
        case .leaveFailed:
            return "Failed to leave event"
        case .reviewFailed:
            return "Failed to add review"
        case .commentFailed:
            return "Failed to add comment"
        case .upgradeFailed:
            return "Failed to upgrade event visibility"
        }
    }
} 